### ISSUE TITLE
Encrypting requests and responses

### ISSUE THREAD

**Issue by wrygiel** [erasmus-without-paper/ewp-specs-architecture#26]
We already designed HTTP Signatures for request/response authentication. As [agreed in Porto](https://github.com/erasmus-without-paper/ewp-specs-architecture/issues/22), we still need to design request/response encryption. But what standard(s) do we want to use for that?

### Solution 1: `aes128gcm`

We could use `Content-Encoding: aes128gcm`, as defined in [RFC 8188](https://tools.ietf.org/html/rfc8188). This is well documented and covers what we need (and more).

We would need to extend it to somehow transmit and RSA-encrypted AES key. This can be done by a custom HTTP header, or by utilizing the `keyid` part of the RFC 8188's binary header.

```http
Content-Encoding: aes128gcm

<header> <encrypted-record> <encrypted-record> <encrypted-record>
```

Pros:

* We follow a well-documented standard.

Cons:

* It's quite complex, and we probably don't need most of its features (authentication, streaming, range-queries).
* I couldn't find existing libraries that would directly refer to this spec.

### Solution 2: Pick one from Java's [Cipher class](https://docs.oracle.com/javase/7/docs/api/javax/crypto/Cipher.html)

We simply pick one of the default encryption algorithms provided by the Java's `Cipher` class. This approach doesn't share some of the cooler properties of Solution 1, but we probably won't need those in EWP anyway.

It seems that `AES/CBC/PKCS5Padding` fits all what we need. As in Solution 1, we need to transmit and RSA-encrypted AES key (e.g. via a custom HTTP header). We will also need to transmit the salt (initialization vector), in order to allow for a single AES key to be used multiple times (but only for the same recipient, of course).

```http
Content-Encoding: ewp-simple-aes
Recipient-Public-Key-Id: 035013774f596e1887344bda8a06462030759592355db3f51c8498ffd0b18add
Encrypted-Symmetric-Key: IAaRLMm4AGvuGrxoElibijAFYT0YIu6Gaew0vS+/onOGjiFhyZ6LcmNMo/LNrCmT
    8PCZCjBZqVLje49M6sONeT1Hg7cOfojiO8i/C6P5MMVgr/czhQ5A/96KBIYe6V2+
    RUQItSbPtsKcaLTjKAuX5AjSZSiP3yp7x7TzmRXFQ0xZh5oE7OSr7Ep3P0MGjEhL
    LNnv+b2BbY5foZBH69QJOAlIcRJoFc4bdeXO97pCgJnhZbu/Hsrtak0xuoMSQooI
    gmAii8VjolgZlNCCQ8/20931exqHqDNNaOdsISZx7x0FUjrfInF1RJ1X0RYxYdnr
    wsnWowZAePkYMTmnu3QPGQ==

<initialization-vector> <encrypted-content>
```

Pros:

* Easier to implement.

Cons:

* Doesn't follow any standard.

@erasmus-without-paper/all-members, any thoughts?

**Comment by wrygiel** [comment]
### Solution 2b

Same as solution 2, but all the binary content will be included in the body. No extra headers will be used.

```http
Content-Encoding: ewp-simple-aes

<recipient-public-key-id>
<encrypted-symmetric-key-length>
<encrypted-symmetric-key>
<initialization-vector>
<encrypted-content>
```

### Solution 2c

The exact opposite of solution 2b. All information except the encrypted content will be included in HTTP headers.

```http
Content-Encoding: ewp-simple-aes
Recipient-Public-Key-Id: 035013774f596e1887344bda8a06462030759592355db3f51c8498ffd0b18add
Encrypted-Symmetric-Key: IAaRLMm4AGvuGrxoElibijAFYT0YIu6Gaew0vS+/onOGjiFhyZ6LcmNMo/LNrCmT
    8PCZCjBZqVLje49M6sONeT1Hg7cOfojiO8i/C6P5MMVgr/czhQ5A/96KBIYe6V2+
    RUQItSbPtsKcaLTjKAuX5AjSZSiP3yp7x7TzmRXFQ0xZh5oE7OSr7Ep3P0MGjEhL
    LNnv+b2BbY5foZBH69QJOAlIcRJoFc4bdeXO97pCgJnhZbu/Hsrtak0xuoMSQooI
    gmAii8VjolgZlNCCQ8/20931exqHqDNNaOdsISZx7x0FUjrfInF1RJ1X0RYxYdnr
    wsnWowZAePkYMTmnu3QPGQ==
Initialization-Vector: I1BsxtFttlv3u_Oo94xnmw

<encrypted-content>
```

**Comment by wrygiel** [comment]
## Example for Solution 2b

This is an example of an encrypted message formatted in the way proposed by Solution 2b. It it in fact binary. Below, all values are encoded in Base64, for presentation reasons.

```
A1ATd09ZbhiHNEvaigZGIDB1lZI1XbP1HISY/9Cxit0AAAEAlbDjuw2fBWN74G5a
S8FUnNgpBos8OcqM20hqJmUijtZuTQLCsvR8CTPEXAhH2rEMMWdkCK6nb++D2qkI
kODuHaL0WVZ+bI6S+PYRFWKiGQ49yUseeuANH/vDMpzcLRHh3i+epF3AFuSw8SIO
zf3hytTgYOeW7iSdfffwr/GVYApjW2sQ1E2oMnrN2NdHux9hSKf9Wt5kBwLEXjUZ
xsf6EY2m1o1gV2vEAA8Lqp/Ic4N3WfMVUG6eElWP3s2AMX4n0ioHki+kQSVzmM5c
XNF8WbAfUHJ9H59poU0WMRFGohoN9M14Stw/lV6GPen31ePxe6mJRM02L2lhx2pa
WTUo5zBBeVyd6KVIje3PVSNX9w7SAbJOozSL0NFPdN0ScQY0t7Lo5zDRptsoHijl
ii/0JQ==
```

Recipients should be able to decrypt this message with the following RSA key-pair:

```
// public
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAl8UYbSjdXHIkDCzbO4Rn
9c/kb4xm3e2A3WSWQ/WtoTqrEy9/nJNe7ehP9gaUrbUf9KCMF0fHypQ6oVasPr/6
zCiHV0LDwiImWbbo+9zehT4foahcdRqPqfEVeRJgMr7hfvcrvtzFO7yFLHkYkt4r
4MLy3GIkSys2arr3vs8V/JY826qTlIi4tLifeeY9I6adMDXPtAoHr9er2XdOt4LW
uLOxqHpddrnuSJhnlY0I1mz8b9LCoN15gwW9d1G/H1KFmIkWu0ty205XUW5c/YRX
xjK1WZcXlfobTyVyeeK2yeEeXdKRm0zuIlBruij3VZ+/dmPXdtn0zGlStOS5UhjS
NQIDAQAB

// private
MIIEvgIBADANBgkqhkiG9w0BAQEFAASCBKgwggSkAgEAAoIBAQCXxRhtKN1cciQM
LNs7hGf1z+RvjGbd7YDdZJZD9a2hOqsTL3+ck17t6E/2BpSttR/0oIwXR8fKlDqh
Vqw+v/rMKIdXQsPCIiZZtuj73N6FPh+hqFx1Go+p8RV5EmAyvuF+9yu+3MU7vIUs
eRiS3ivgwvLcYiRLKzZquve+zxX8ljzbqpOUiLi0uJ955j0jpp0wNc+0Cgev16vZ
d063gta4s7Goel12ue5ImGeVjQjWbPxv0sKg3XmDBb13Ub8fUoWYiRa7S3LbTldR
blz9hFfGMrVZlxeV+htPJXJ54rbJ4R5d0pGbTO4iUGu6KPdVn792Y9d22fTMaVK0
5LlSGNI1AgMBAAECggEANsqlIuOZ5wIeGXcoPrhyf7/qDIt3p69S0pq51Rcg9BAm
Kur++xwJLYKtO3jsvDmjq8E6Uj1L18rjz9Nmo9DTTlljYxFrcu65QbJTMnpuq1Pe
P5J0rqJEM2oiAm+r4yYeaqP5WxKA8iwBOCkPwhYLaT14SC/2Qlz7bFTLlEtW+LUE
b2w9InThtEvIoNK2+n5BSRrjXH46dvJX28dsWft2f5o3+PX8L6z0qNOyKKgOHbBl
w37IbV78GhdI0dLFEGEcMkdV+p2bqChVxiu3S8G/0fwjl6HwhUYD5OOWg3ncWllB
9rLdJNA2zYFfnwsycrgEh1vk0zp0S9cbuoOyc6h+AQKBgQDm6aEHVZ3c+ZIB5jwg
qX9tK2Zr5bdyptqpAfXgDf9Sj34Zg5+0pnA7v8zRlg2rqwE50XRSEiVxpBDdMswR
e30SioN1zHeabROuSAd4EHMkrluSmcObeP3LcStj5D0VAmbijAgYnRKiShGRvUSV
s42pMyxHo+RYTiXk0/tg3hrelQKBgQCoQkVPqw1Dlv8BKDU284ogqVCTNDqz96rs
GNKUmnRR4UVEJZGXOVvfDMt3kYr56QCmDhyGf6ny6W/YgjNJHDbz3LAl9DLc/8d4
y8FgXaBqwPpvW2aBL3tea8Raa7+D3t/Z68sJduueJO921E72n+xa6TvYModtE7cI
wmfMcy5dIQKBgQCO8Mbe3HAJj3CDvnswGNypvrj7R8uErKclAfKrjN6lw+/iaWle
kb1eLz/h6cpynzv2B6PC/jqxm0dZNo2+sLve02HHdRgAv070juAYwc4VQd2r5YWB
46bv3hFnF618KO15hgeo/OrBDarMleYz6V9jAyuA+YJr64xnl5XABB2L9QKBgDzl
Vz6BMti+gmZBzhioRdqSTNYp9gECZvrx9OzRhb3IoRAL5MhtewGcGNuackkGejSf
MNXAyJpgwBkE7ljMfFsACUSDQBFaBTCD1eXxnMhmNX0uAEhLDgRbToJHMtYgSLYP
L7mqL5ZZ2c0RA88gjCNO/Fi/2OGyW/Ewou6M1T/hAoGBAJq1J6wOR05lShFYwyev
0rxGeZCx8bpWBEXwuzPJFZCQxyym9ZexcArI/74a2Fb4EsgxnwORksuBVJDZUG7a
L05xiW1MJrcHyh9DDo47T2Qo5k4NngXNBmOhjbV6cmzJBaKY9j3B/bJmWhILBiSe
GLYpa/SUIFhQuMLqwb/2pKSa
```


### Intermediate values

#### Recipient key ID (32 bytes)

This is the SHA-256 fingerprint of the recipient's public RSA key.

```
// Hex-encoded
035013774f596e1887344bda8a06462030759592355db3f51c8498ffd0b18add

// Base64-encoded
A1ATd09ZbhiHNEvaigZGIDB1lZI1XbP1HISY/9Cxit0=
```


#### AES key

First, there's a 4-byte long integer N (big-endian; probably we'll replace it with 2-byte short in the final spec, for security reasons), followed by N bytes with the encrypted `aesKey`. In this example `N` is `256` (because we have used a 2048-bit RSA key for encryption). The encrypted content is:

```
lbDjuw2fBWN74G5aS8FUnNgpBos8OcqM20hqJmUijtZuTQLCsvR8CTPEXAhH2rEM
MWdkCK6nb++D2qkIkODuHaL0WVZ+bI6S+PYRFWKiGQ49yUseeuANH/vDMpzcLRHh
3i+epF3AFuSw8SIOzf3hytTgYOeW7iSdfffwr/GVYApjW2sQ1E2oMnrN2NdHux9h
SKf9Wt5kBwLEXjUZxsf6EY2m1o1gV2vEAA8Lqp/Ic4N3WfMVUG6eElWP3s2AMX4n
0ioHki+kQSVzmM5cXNF8WbAfUHJ9H59poU0WMRFGohoN9M14Stw/lV6GPen31ePx
e6mJRM02L2lhx2paWTUo5w==
```

The decryted AES key content is only 16 bytes (128 bits, because we use AES 128):

```
PjdFvIrTJfCJDdOtpqvDmw==
```


#### Salt (IV)

Next, there's the random 16-byte initialization vector for the AES. In our example the vector is:

```
MEF5XJ3opUiN7c9VI1f3Dg==
```


#### Encrypted content

The actual AES-encrypted content starts at offset 308, and is:

```
0gGyTqM0i9DRT3TdEnEGNLey6Ocw0abbKB4o5Yov9CU=
```


#### Decrypted content

```
This is a secret
```

**Comment by wrygiel** [comment]
Is anyone against choosing this solution for our EWP encryption?

**Comment by mpuzar** [comment]
The funding for the project is finished. You will have to wait with this until January when EWP2 has started.

**Comment by vgiralt** [comment]
I have a tendency to prefer RFC specifications. Agreed that RFC 8188 might have things we do not need, but it is a standard.

On the library part, it strikes me, as RFC approval requires reference running code.

Finally, about soluitons 2x, why are we transmitting symmetric keys? Do existing public key based encryption methods not serve our purposes for any reason that escapes me?

**Comment by wrygiel** [comment]
> On the library part, it strikes me, as RFC approval requires reference running code.

If so, then how can we find this reference code?

> Finally, about soluitons 2x, why are we transmitting symmetric keys? Do existing public key based encryption methods not serve our purposes for any reason that escapes me?

What existing encryption methods do you mean? Note, that Solution 1 also uses symmetric keys. As does TLS, and every other reasonable encryption solution I found.

**Comment by vgiralt** [comment]
My excuses Wojciech, can't learn not to respond too fast in the wee hours.
You are right on the exchange of the symmetric keys, it was the time of the day. The encryption methods I was thinking about were OpenPGP message format [RFC4880](https://tools.ietf.org/html/rfc4880), that in section 2.1 clearly states how the symmetric key is exchanged, and S/MIME (heavier reading with lots of ASN.1, [RFC3369, section 12](https://tools.ietf.org/html/rfc3369#section-12)).

On the reference code side, I have not had much luck, like yourself, although I have found a reference to webpush in an [IETF blog](https://www.ietf.org/blog/2017/10/privacy-and-trustworthiness-for-web-notifications/) that has led me to the [W3C Webpush spec](https://w3c.github.io/push-api/), through the [Can I use ___? site](https://caniuse.com/#feat=push-api). Really not much, but cannot dedicate more time to this right now, sorry :-(

**Comment by mpuzar** [comment]
At a first glance, I could live with any of the proposed solutions. Agreed, using an RFC is nice, and this one is well described, but maybe an overkill for our case. Of the solutions 2x, I like 2c best (where most of the necessary information is in the headers). Do we expect the encrypted messages to be stored? If so, it might be better to have everything in the body to avoid having to store headers separately, but I am not sure this is a case for us.

Regarding the symmetric keys, since we don't operate with sessions, I guess we can (should?) use a new key for each message? That might make some implementation details easier.

In any case, I will be happy to make our internal reference implementation of whatever we decide to go for.

**Comment by wrygiel** [comment]
> Agreed, using an RFC is nice, and this one is well described, but maybe an overkill for our case.

I think so too.

> Of the solutions 2x, I like 2c best (where most of the necessary information is in the headers). Do we expect the encrypted messages to be stored? If so, it might be better to have everything in the body to avoid having to store headers separately, but I am not sure this is a case for us.

- 2b is not easily extendable (the format is fixed), but is more self-contained (the body contains everything needed to identify the recipient and allow him to decrypt the message).
- 2c is more extensible (extra headers can be added later on, if they're needed), but depends on extra headers (so we need to sign them, and verify that they're signed, etc.). Also, there's a chance of [reaching the header length limit](http://httpd.apache.org/docs/2.2/mod/core.html#limitrequestfieldsize) (depends on the webserver configuration).

I find 2b slightly less troublesome. The lack of extensibility doesn't seem to be a real problem, because we can always introduce a new `Content-Encoding` value, if we ever need to extend it. But both 2b and 2c are fine for me.

> Regarding the symmetric keys, since we don't operate with sessions, I guess we can (should?) use a new key for each message? That might make some implementation details easier.

Yes, it's perfectly valid to generate a new AES key for each message. But there's also no need to require that. Generating new AES key is very fast (~1000000/sec), but encrypting and decrypting it with RSA is slower (~5000/sec, for 2k RSA key). If the sender tries to reuse the same AES key, then there's a way to optimize the process by caching (both parties keeping proper mappings with encrypted and decrypted AES keys in memory, without the need to encrypt/decrypt it every time).

But it's also true that requiring a new AES key per each message is a bit simpler, because we won't need salt (the IV record) in this case.

**Comment by BavoNootaert** [comment]
Solution 1 looks like quite some work to implement, lacking an existing library.

For solution 2, why do you propose CBC instead of GCM as in solution 1? E.g. XML-Enc only keeps AES-CBC for backwards compatibility, but encourages AES-GCM.

I prefer 2b over 2c. The header length may not be a problem: all HTTP Signature headers together are about 1KB, and the examples above are 0.5KB. Servers like Apache typically allow 8KB per header.
Signing the extra headers, however, would make the HTTP Signature implementation more complex, especially if we later add new headers and deprecate old ones.
As for extensibility in 2b: if we are designing the encoding ourselves, we don't _have_ to just concatenate the fields as in the example above. We may e.g. separate them by new-line characters, and prefix them with a name and a colon, thus allowing for future extensions.

**Comment by wrygiel** [comment]
> For solution 2, why do you propose CBC instead of GCM as in solution 1? E.g. XML-Enc only keeps AES-CBC for backwards compatibility, but encourages AES-GCM.

GCM is often encouraged because, besides the encryption, it also handles integrity. However, we have already chosen `httpsig` for handling integrity, so GCM seems redundant in our case.

**Comment by BavoNootaert** [comment]
If I remember correctly, we have httpsig for the case where, for some reason, the TLS endpoint cannot be configured to use (self-signed) client certificates. Encryption is for the case where the TLS endpoint cannot be trusted. Maybe we should first specify how the manifest and registry should be handled in this case.

**Comment by wrygiel** [comment]
> Maybe we should first specify how the manifest and registry should be handled in this case.

I'm not sure if I understand what you mean, but this seems to be a separate issue, I think. Perhaps you could start a new thread for that?

**Comment by lvalente** [comment]
Dear all,
After a internal discussion in my team, we suggest to use OpenPGP and S/MIME, following the @vgiralt  suggestion.
It's the simplest and strongest way to do it.

**Comment by wrygiel** [comment]
> we suggest to use OpenPGP and S/MIME

We have already started writing the specs for solution 2b, because most of the previous commenters seemed to agree that it's good enough.

I won't have time to investigate other proposals any further, but our "v2 security" framework is built in such a manner that many independant alternatives can be proposed, specified, used and promoted by partners which prefer using them.

@lvalente, if you believe that your "OpenPGP and S/MIME" idea is better, then don't hesitate to create your own repository and prepare detailed specification of how this would work, e.g. explain exactly how OpenPGP is supposed to work together with S/MIME (as it seems you propose to use both at once), and how PGP (sub)keys should work with our catalogue's `<rsa-public-key>` elements, etc.

**Comment by BavoNootaert** [comment]
We should try keep the number of supported encryption methods limited. The registry indeed supports using several methods, but we think this should only be used when a method becomes obsolete. Then, during the transition phase, we can _temporarily_ use two methods. With several methods available from the start, requiring all methods to be implemented may be too big a hurdle for entry into EWP, and when we do not require this, the EWP network may become too fragmented.
S/MIME and OpenPGP have the benefit that they are standards, and are supported by libraries, e.g. (for Java) Bouncy Castle. As we have discussed before for authentication, using established standards should be preferred.
Awaiting a complete specification of how it should work (and a specification of the other proposed methods), I think we should keep the option open of adopting it as the main encryption method.

Continuing the cbc versus gcm comment I made earlier, I opened an issue in the [new repository](https://github.com/erasmus-without-paper/ewp-specs-sec-rsa-aes128cbc).

**Comment by wrygiel** [comment]
> With several methods available from the start, requiring all methods to be implemented may be too big a hurdle for entry into EWP, and when we do not require this, the EWP network may become too fragmented.

I don't think we are in a position to require that much (note, that we have already tried that when we required all partners to trust TLS). Of course, fragmentation is one of the possible side-effects of not requiring things. But hopefully, as the network grows, certain standards will become self-apparent, and fragmentation won't be a big issue.

> I think we should keep the option open

I agree. If someone else has a time to specify a different solution, and - in time - it turns out that all partners prefer it - then the one we specify now can be deprecated. Note, that the opposite is also possible - for example, the majority of partners may start trusting TLS again, and we will deprecate the new solutions in favor of the old ones.

BTW - implementing a Java library for solution 1 also seems a valid option (it's more work, but the code would be somewhat reusable). Perhaps someone would be willing to specify and implement Solution 1?

> Continuing the cbc versus gcm comment I made earlier, I opened an issue

I have switched the specs to GCM in version 0.2.0.
